stdlib/Test/src/GenericDimensionfuls.jl

```diff
@@ -0,0 +1,118 @@
+
+module GenericDimensionfuls
+
+export GenericDimensionful
+
+# Here we implement a minimal dimensionful type GenericDimensionful, which is used
+# to test dimensional correctness of various functions in Base.
+
+"""
+The `GenericDimensionful` type can be used to test that numerical code works with
+dimensionful quantities.   Construct a dimensionful quantity with `GenericDimensionful(x)`
+(or `GenericDimensionful{p}(x)` for unitsᵖ).  Arithmetic operations for `GenericDimensionful`
+are defined so that you can only combine quantities in ways that respect the units.
+"""
+struct GenericDimensionful{p,T<:Number} <: Number
+    val::T
+    function GenericDimensionful{p,T}(v::Number) where {p,T}
+        p isa Real || throw(DomainError("units must have real powers"))
+        new(v)
+    end
+end
+GenericDimensionful(x::T) where {T<:Number} = GenericDimensionful{1,T}(x)
+GenericDimensionful(x::GenericDimensionful) = x
+(::Type{T})(x::GenericDimensionful) where {T<:Number} = T(x.val)::T
+GenericDimensionful{p}(v::Number) where {p} = GenericDimensionful{p,typeof(v)}(v)
+GenericDimensionful{p}(x::GenericDimensionful{q}) where {p,q} = GenericDimensionful{p,typeof(x.val)}(x.val)
+GenericDimensionful{p,T}(x::GenericDimensionful{q}) where {T,p,q} = GenericDimensionful{p,T}(T(x.val))
+
+Base.convert(::Type{GenericDimensionful{p}}, x::GenericDimensionful{p}) where {p} = x
+Base.convert(::Type{GenericDimensionful{p,T}}, x::GenericDimensionful{p}) where {p,T} = GenericDimensionful{p,T}(x)
+Base.convert(::Type{GenericDimensionful{0}}, x::Union{Real,Complex}) = GenericDimensionful{0}(x)
+Base.convert(::Type{GenericDimensionful{0,T}}, x::Union{Real,Complex}) where {T} = GenericDimensionful{0}(convert(T, x))
+Base.convert(D::Type{GenericDimensionful{p}}, x::Number) where {p} = throw(DimensionMismatch("dimension mismatch between $D and $(typeof(x))"))
+Base.convert(D::Type{GenericDimensionful{p,T}}, x::Number) where {p,T} = throw(DimensionMismatch("dimension mismatch between $D and $(typeof(x))"))
+
+# need separate method to handle case where p==q but p!==q, e.g. p=1.0 and q=1
+Base.convert(D::Type{GenericDimensionful{p}}, x::GenericDimensionful{q}) where {p,q} =
+    p == q ? GenericDimensionful{p}(x.val) : throw(DimensionMismatch("dimension mismatch between $D and $(typeof(x))"))
+Base.convert(D::Type{GenericDimensionful{p,T}}, x::GenericDimensionful{q}) where {p,q,T} =
+    p == q ? GenericDimensionful{p,T}(x.val) : throw(DimensionMismatch("dimension mismatch between $D and $(typeof(x))"))
+
+Base.promote_type(::Type{GenericDimensionful{p,T}}, ::Type{GenericDimensionful{p,S}}) where {p,T,S} =
+    (Base.@_pure_meta; GenericDimensionful{p,promote_type(T,S)})
+Base.promote_type(A::Type{GenericDimensionful{p,T}}, B::Type{GenericDimensionful{q,S}}) where {p,q,T,S} =
+    (Base.@_pure_meta; p == q ? GenericDimensionful{p,promote_type(T,S)} : throw(DimensionMismatch("dimension mismatch between $A and $B")))
```

Comment:
```suggestion
@generated Base.promote_rule(::Type{GenericDimensionful{p,T}}, ::Type{GenericDimensionful{q,S}}) where {p,q,T,S} =
    p === q ? :(GenericDimensionful{p, promote_type(T,S)}) :
    p == q  ? :(GenericDimensionful{$(canonical_p(p)), promote_type(T,S)}) : :(Union{})
```
`promote_type` shouldn’t throw, the error handling happens in `promote` (at least that’s how the fallback implementation works).

Reply:
Done, thanks.